driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -96,108 +113,137 @@ public void connect() {
         monitorThread.interrupt();
+        roundTripTimeMonitor.close();
+        roundTripTimeMonitorThread.interrupt();
+    }
+
+    @Override
+    public void cancelCurrentCheck() {
+        monitor.cancelCurrentCheck();
     }
 
     class ServerMonitorRunnable implements Runnable {
-        private final ExponentiallyWeightedMovingAverage averageRoundTripTime = new ExponentiallyWeightedMovingAverage(0.2);
+        private volatile InternalConnection connection = null;
+        private volatile boolean currentCheckCancelled;
+
+        void close() {
+            InternalConnection connection = this.connection;
+            if (connection != null) {
+                connection.close();
+            }
+        }
 
         @Override
-        @SuppressWarnings("unchecked")
         public synchronized void run() {
-            InternalConnection connection = null;
-            try {
-                ServerDescription currentServerDescription = getConnectingServerDescription(null);
-                while (!isClosed) {
-                    ServerDescription previousServerDescription = currentServerDescription;
-                    try {
-                        if (connection == null) {
-                            connection = internalConnectionFactory.create(serverId);
-                            try {
-                                connection.open();
-                            } catch (Throwable t) {
-                                connection = null;
-                                throw t;
-                            }
-                        }
-                        try {
-                            currentServerDescription = lookupServerDescription(connection);
-                        } catch (MongoSocketException e) {
-                            connectionPool.invalidate();
-                            connection.close();
-                            connection = null;
-                            connection = internalConnectionFactory.create(serverId);
-                            try {
-                                connection.open();
-                            } catch (Throwable t) {
-                                connection = null;
-                                throw t;
-                            }
-                            try {
-                                currentServerDescription = lookupServerDescription(connection);
-                            } catch (MongoSocketException e1) {
-                                connection.close();
-                                connection = null;
-                                throw e1;
-                            }
-                        }
-                    } catch (Throwable t) {
-                        averageRoundTripTime.reset();
-                        currentServerDescription = getConnectingServerDescription(t);
-                    }
+            ServerDescription currentServerDescription = getConnectingServerDescription(null);
```

Comment:
Compare this to the new pseudocode in the specification.  It now follows it much more closely, which simplified the logic a bit.